aten/src/ATen/native/cuda/TensorFactories.cuh

```diff
@@ -0,0 +1,55 @@
+#include <ATen/CUDAGeneratorImpl.h>
+#include <ATen/cuda/detail/UnpackRaw.cuh>
+#include <ATen/Utils.h>
+
+#include <curand.h>
+#include <curand_kernel.h>
+#include <curand_philox4x32_x.h>
+
+namespace {
+
+// See note [Algorithm of randperm]
+template<typename T, typename scalar_t>
+__global__ void randperm_handle_duplicate_keys_kernel(T *keys, scalar_t *data, T mask, int n, at::PhiloxCudaState philox_args) {
+  int tid = threadIdx.x + blockDim.x * blockIdx.x;
+
+  // find the beginning of islands
+  if (tid >= n - 1) return;  // out of range
+  if ((keys[tid] & mask) != (keys[tid + 1] & mask)) return;  // not in an island
+  if (tid != 0 && (keys[tid] & mask) == (keys[tid - 1] & mask)) return;  // not the beginning of an island
+
+  // find the size of islands
+  int island_size = 0;
+  while ((keys[tid + ++island_size] & mask) == (keys[tid] & mask));
+
+  // do random permutation inside each island.
+  data += tid;
+  auto seeds = at::cuda::philox::unpack(philox_args);
+  curandStatePhilox4_32_10_t state;
+  curand_init(std::get<0>(seeds), tid, std::get<1>(seeds), &state);
+  for (int i = island_size - 1; i > 0; i--) {
+    unsigned int r = curand(&state) % (i + 1);
+    if (i != r) {
+      scalar_t tmp = data[i];
```

Comment:
Shouldn't you be accessing based on tid here?

Reply:
There is a
```C++
data += tid;
```
in https://github.com/pytorch/pytorch/pull/54113/files/bda1c838d1b85a54562ad85fa508a3a434f29589#diff-db9fbcdb96e426b3b546a6fe794fb864e74cebb1cf31ec70ccd97b7db364ad37R26